ddtrace/contrib/unittest/patch.py

```diff
@@ -103,85 +278,328 @@ def patch():
-            span_type=SpanTypes.TEST,
-        ) as span:
-            span.set_tag_str(_EVENT_TYPE, SpanTypes.TEST)
+def add_skip_test_wrapper(func, instance, args, kwargs):
+    if _is_valid_result(instance, args):
+        _set_test_span_status(test_item=args[0], skip_reason=_extract_test_reason(args), status=test.Status.SKIP.value)
 
-            span.set_tag_str(COMPONENT, COMPONENT_VALUE)
-            span.set_tag_str(SPAN_KIND, KIND)
+    return func(*args, **kwargs)
 
-            span.set_tag_str(test.FRAMEWORK, FRAMEWORK)
-            span.set_tag_str(test.TYPE, SpanTypes.TEST)
-            suite_name = _extract_suite_name_from_test_method(instance)
-            span.set_tag_str(test.NAME, _extract_test_method_name(instance))
-            span.set_tag_str(test.SUITE, suite_name)
-            span.set_tag_str(test.MODULE, _extract_module_name_from_test_method(instance))
 
-            span.set_tag_str(test.STATUS, test.Status.FAIL.value)
-            span.set_tag_str(test.CLASS_HIERARCHY, suite_name)
-            _CIVisibility.set_codeowners_of(_extract_test_file_name(instance), span=span)
+def add_xpass_test_wrapper(func, instance, args, kwargs):
+    if _is_valid_result(instance, args):
+        _set_test_xpass_xfail_result(test_item=args[0], result=test.Status.XPASS.value)
 
-            _store_span(instance, span)
-            result = func(*args, **kwargs)
-            return result
+    return func(*args, **kwargs)
 
+
+def handle_test_wrapper(func, instance, args, kwargs):
+    if _is_valid_test_call(kwargs) and _is_test(instance) and hasattr(_CIVisibility, "_unittest_data"):
+        test_suite_name = _extract_suite_name_from_test_method(instance)
+        test_name = _extract_test_method_name(instance)
+        test_module_path = _extract_module_file_path(instance)
+        test_module_suite_path = _generate_module_suite_path(test_module_path, test_suite_name)
+        test_suite_span = _extract_suite_span(test_module_suite_path)
+        test_module_span = _extract_module_span(test_module_path)
+        if test_module_span is None and test_module_path in _CIVisibility._unittest_data["modules"]:
+            test_module_span = _start_test_module_span(instance)
+            _CIVisibility._unittest_data["modules"][test_module_path]["module_span"] = test_module_span
+        if test_suite_span is None and test_module_suite_path in _CIVisibility._unittest_data["suites"]:
+            test_suite_span = _start_test_suite_span(instance)
+            _CIVisibility._unittest_data["suites"][test_module_suite_path]["suite_span"] = test_suite_span
+        if not test_module_span or not test_suite_span:
+            log.debug("Suite and/or module span not found for test: %s", test_name)
+            return func(*args, **kwargs)
+        span = _start_test_span(instance, test_suite_span)
+        result = func(*args, **kwargs)
+        _update_status_item(test_suite_span, span.get_tag(test.STATUS))
+        span.finish()
```

Comment:
We should use a context manager here. This will ensure span.finish() is always called.
```suggestion
        with _start_test_span(instance, test_suite_span) as span:
          result = func(*args, **kwargs)
          _update_status_item(test_suite_span, span.get_tag(test.STATUS))